<H1>PROJECT: UK Road Accidents</H1>

<h2>Analyst: Ramos, Alyssa Sophia A.</h2>

In [3]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [5]:
accidents= pd.read_csv("uk_datasets\\accident_data.csv")

In [6]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [7]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [8]:
accidents['Latitude'] = accidents['Latitude'].fillna(accidents['Latitude'].mode()[0])
accidents['Longitude'] = accidents['Longitude'].fillna(accidents['Longitude'].mode()[0])
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].fillna(accidents['Road_Surface_Conditions'].mode()[0])
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].fillna(accidents['Urban_or_Rural_Area'].mode()[0])
accidents['Road_Type'] = accidents['Road_Type'].fillna(accidents['Road_Type'].mode()[0])
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].fillna(accidents['Weather_Conditions'].mode()[0])

In [9]:
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [10]:
accidents.info

<bound method DataFrame.info of                 Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1  

In [11]:
accidents['Accident Date'] = pd.to_datetime(accidents['Accident Date'], dayfirst = True, errors = "coerce")

In [12]:
accidents.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

<h1>Extracting Date Information using Pandas Date Time</h1>

In [13]:
accidents['Year']= accidents ['Accident Date'].dt.year
accidents['Month']= accidents ['Accident Date'].dt.month
accidents['Day']= accidents ['Accident Date'].dt.day
accidents['DayOfWeek']= accidents ['Accident Date'].dt.dayofweek

<h1>Questions:</h1>

<h1>1. Average number of vehicles involves per accident</h1>

In [14]:
accidents["Number_of_Vehicles"].mean()

np.float64(1.8312554205597575)

<h3>Insight:On average, around 2 vehicles are involved per accident. </h3>

<h1>2. How many accidents happened in urban vs rural areas?</h1>

In [15]:
accidents['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h3>Insight: Most accidents occur in urban areas compared to rural ones </h3>

<h1>3. Are severe accidents linked with more vehicles?</h1>

In [16]:
accidents.groupby("Accident_Severity")['Number_of_Vehicles'].mean()

Accident_Severity
Fatal      1.786976
Serious    1.678327
Slight     1.855864
Name: Number_of_Vehicles, dtype: float64

<h3>Insight: Fatal accidents often involve more vehicles</h3>

<h1>4. Which 10 districts have the highest fatal accident counts?</h1>

In [17]:
accidents[accidents["Accident_Severity"] == "Fatal"]["District Area"].value_counts().head(10)

District Area
Birmingham                  105
Leeds                        93
Highland                     88
East Riding of Yorkshire     85
Bradford                     71
Aberdeenshire                66
Powys                        59
Wakefield                    56
Doncaster                    56
Herefordshire, County of     51
Name: count, dtype: int64

<h3>Insight: This shows which 10 district have the highest fatal accident counts. Birmingham have the highest fatal accident counts, while Hereforshire have the lowest fatal accident counts</h3>

<h1>5. Which 5 districts have the lowest number of accidents?</h1>

In [18]:
accidents["District Area"].value_counts().tail(5)

District Area
Berwick-upon-Tweed    153
Teesdale              142
Shetland Islands      133
Orkney Islands        117
Clackmannanshire       91
Name: count, dtype: int64

<h3>Insight: Some rural or less populated districts report the fewest accidents</h3>

<h1>6. Distributions of accidents by day of the week</h1>

In [19]:
accidents["Accident Date"].dt.day_name().value_counts()

Accident Date
Saturday     107178
Wednesday     99558
Thursday      99511
Friday        97900
Tuesday       94550
Sunday        89302
Monday        72680
Name: count, dtype: int64

<h3>Insight: Weekdays, especially Friday, often have the most accidenets</h3>

<h1>7.Which months have the most fatal accidents?</h1>

In [20]:
accidents[accidents['Accident_Severity']=='Fatal']['Accident Date'].dt.month.value_counts().sort_index()

Accident Date
1     740
2     617
3     681
4     697
5     760
6     673
7     753
8     801
9     723
10    798
11    747
12    671
Name: count, dtype: int64

<h3>Insight: August have the most recorded fatal accident,meanwhile February have the fewest.</h3>

<h1>8. Which road surface conditions caused more cars to have accidents?</h1>

In [21]:
accidents[accidents['Vehicle_Type']=="Car"]['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     337860
Wet or damp             140764
Frost or ice             14108
Snow                      4483
Flood over 3cm. deep       777
Name: count, dtype: int64

<h3>Insights: Most accidents happened on dry roads, followed by wet roads</h3>

<h1>9. Do serious accidents involve more casualties than fatal ones (on average)?</h1>

In [22]:
accidents.groupby("Accident_Severity")["Number_of_Casualties"].mean()

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h3>Insights: Yes, serious accidents had more people injured per crash than fatal ones</h3>

<h1> 10. which vehicle type has the highest average casualties per crash?</h1>

In [23]:
accidents.groupby('Vehicle_Type')['Number_of_Casualties'].mean().sort_values(ascending=False).head(10)

Vehicle_Type
Data missing or out of range            1.500000
Pedal cycle                             1.370558
Taxi/Private hire car                   1.368663
Goods over 3.5t. and under 7.5t         1.362861
Motorcycle over 500cc                   1.359434
Car                                     1.358841
Van / Goods 3.5 tonnes mgw or under     1.354537
Goods 7.5 tonnes mgw and over           1.351881
Bus or coach (17 or more pass seats)    1.349216
Minibus (8 - 16 passenger seats)        1.345648
Name: Number_of_Casualties, dtype: float64

<h3>Insights: Buses and coaches had the highest casualties per accident, since they carry more passengers</h3>

<h1>11. How many accidents happend in each year of the dataset?</h1>

In [24]:
accidents['Accident Date'].dt.year.value_counts().sort_index()

Accident Date
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h3>Insights: the dataset shows accidents for several years, with a general decrease over time</h3>

<h1>12.Do accidents in wet road conditions cause more casualties?</h1>

In [25]:
accidents.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean()

Road_Surface_Conditions
Dry                     1.334545
Flood over 3cm. deep    1.495575
Frost or ice            1.348815
Snow                    1.359593
Wet or damp             1.411064
Name: Number_of_Casualties, dtype: float64

<h3>Insight: Yes, crashes on wet roads involved slightly more casualties than on dry roads </h3>

<h1>13. Percentage of accidents that are "Fatal"</h1>

In [26]:
(accidents['Accident_Severity']=='Fatal').mean() *100

np.float64(1.3109240644851736)

<h3>Insight: About 1.3% of all accidents were fatal</h3>

<h1>14. which day of the week have the most fatal accident recorded</h1>

In [27]:
accidents[accidents['Accident_Severity']=='Fatal']['Accident Date'].dt.dayofweek.value_counts().sort_index()

Accident Date
0    1385
1    1105
2    1113
3    1097
4    1113
5    1326
6    1522
Name: count, dtype: int64

<h3>Insight: Saturday had the most fatal crashes, followed by Friday</h3>

<h1>15. How many accidents had more than 3 casualties?</h1>

In [28]:
(accidents['Number_of_Casualties']>3).sum()

np.int64(17891)

<h3>Insight: There were 17891 accidents with more than 3 casualties</h3>

<h1>16. Average number of vehicles by accident severity and road type</h1>

In [29]:
accidents.groupby(['Accident_Severity', 'Road_Type'])['Number_of_Vehicles'].mean()

Accident_Severity  Road_Type         
Fatal              Dual carriageway      1.932231
                   One way street        1.357895
                   Roundabout            1.549296
                   Single carriageway    1.759756
                   Slip road             1.571429
Serious            Dual carriageway      1.819428
                   One way street        1.389728
                   Roundabout            1.715689
                   Single carriageway    1.660009
                   Slip road             1.638298
Slight             Dual carriageway      2.037443
                   One way street        1.614701
                   Roundabout            1.919497
                   Single carriageway    1.818013
                   Slip road             1.946874
Name: Number_of_Vehicles, dtype: float64

<h3>Insights: Fatal accidents on highways involved more vehicles than slight accidents on local roads</h3>

<h1>17. Do multi-vehicle accidents lead to more casualties on average?</h1>

In [30]:
accidents.groupby(accidents["Number_of_Vehicles"] > 2) ["Number_of_Casualties"].mean()

Number_of_Vehicles
False    1.305799
True     1.804410
Name: Number_of_Casualties, dtype: float64

<h3>Insight: Yes, accidents with more than 2 vehicles caused more injuries on average</h3>

<h1>18. Top 5 vehicle types most often in fatal accidents</h1>

In [31]:
accidents[accidents["Accident_Severity"]=='Fatal']["Vehicle_Type"].value_counts().head(5)

Vehicle_Type
Car                                     6577
Van / Goods 3.5 tonnes mgw or under      467
Motorcycle over 500cc                    339
Bus or coach (17 or more pass seats)     325
Goods 7.5 tonnes mgw and over            216
Name: count, dtype: int64

<h3>Insights:The top vehicles in fatal accidents were: Cars, Motorcycle, Vans, Trucks and Buses </h3>

<h1>19. Which weather condition has the highest average casualties per accident?</h1>

In [32]:
accidents.groupby("Weather_Conditions")['Number_of_Casualties'].mean().sort_values(ascending=False).head(10)

Weather_Conditions
Fog or mist              1.452948
Snowing + high winds     1.418079
Raining + high winds     1.416641
Raining no high winds    1.408214
Fine + high winds        1.386018
Other                    1.354869
Fine no high winds       1.347397
Snowing no high winds    1.341776
Name: Number_of_Casualties, dtype: float64

<h3>Insight: Foggy conditions and heavy rain showed the highest casualties per crash</h3>

<h1>20. Are accidents with missing road type data more or less severe?</h1>

In [33]:
accidents.groupby(accidents["Road_Type"].isna())['Accident_Severity'].value_counts(normalize=True)

Road_Type  Accident_Severity
False      Slight               0.853366
           Serious              0.133525
           Fatal                0.013109
Name: proportion, dtype: float64

<h3>Insight: Most accidents with missing road type were slight, so they are less severe</h3>

<h1> 21. Which 10 district have the highest average casualties per accident?</h1>

In [34]:
accidents.groupby("District Area")['Number_of_Casualties'].mean().sort_values(ascending=False).head(10)

District Area
Blaeu Gwent          1.693694
West Somerset        1.613990
Derwentside          1.606186
Chester-le-Street    1.603306
Durham               1.580431
Blaenau Gwent        1.579310
Fylde                1.569350
Hambleton            1.568523
Merthyr Tydfil       1.568182
West Lancashire      1.566341
Name: Number_of_Casualties, dtype: float64

<h3>Insight: The highest averages came from busy districts with dense traffic</h3>

<h1>22. How many accidents happened on each day of the month?</h1>

In [35]:
accidents["Accident Date"].dt.day.value_counts().sort_index()

Accident Date
1     22606
2     21815
3     21540
4     21699
5     22409
6     21974
7     21431
8     21914
9     22252
10    22328
11    22503
12    22536
13    21816
14    21474
15    21818
16    21972
17    22117
18    21719
19    21956
20    21838
21    21641
22    21634
23    21997
24    21301
25    19949
26    21168
27    21038
28    20707
29    20138
30    19573
31    11816
Name: count, dtype: int64

<h3>Insight: Accidents are spread across all days, with slight peaks mid-month</h3>

<h1>23. Compare the average casualties between day and night</h1>

In [36]:
accidents.groupby(accidents['Light_Conditions'].str.contains("Daylight"))["Number_of_Casualties"].mean()

Light_Conditions
False    1.403188
True     1.340309
Name: Number_of_Casualties, dtype: float64

<h3>Insight: Night accidents had higher casualties than day accidents</h3>

<h1>24. Which district has the lowest accident density(accidents per year)?</h1>

In [37]:
accidents.groupby(["District Area",accidents["Accident Date"].dt.year]).size().groupby('District Area').mean().sort_values(ascending=False).head(10)

District Area
Birmingham            3372.750000
Leeds                 2224.500000
Manchester            1680.000000
Glasgow City          1647.333333
Bradford              1553.000000
Sheffield             1427.500000
Westminster           1426.500000
Liverpool             1396.750000
Cornwall              1303.000000
Edinburgh, City of    1274.666667
dtype: float64

<h3>Insight: This shows that the district of Edinburgh has the lowest accident density per year</h3>

<h1>25. What percentage of accidents happen in darknesss(all types)? </h1>

In [38]:
accidents["Light_Conditions"].str.contains("Darkness").mean()*100

np.float64(26.60883727195809)

<h3>Insight: About 26.6% of accidents happened in darkness (with or w/o street lights)</h3>